### PR TITLE
Syncing of UsdQt with pixar master

### PR DESCRIPTION
**Pull request by pxmkuruc**

### CODE REVIEW

File: usdQt/hierarchyModel.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+
+class HierarchyStandardModel(HierarchyBaseModel):
+    Name = "Name"
+    Type = "Type"
+    Kind = "Kind"
+
+    NormalColor = QtGui.QBrush(QtGui.QColor(227, 227, 227))
+    InactiveDarker = 200
+
+    ArcsIconPath = 'icons/arcs_2.xpm'
```
Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: They are.  I'll try to get them uploaded in the next push.

**Comment by pxmkuruc** [comment]
Do you have a strong preference for approach?  We don't have cmake_pip internally.

**Comment by chadrik** [comment]
From looking at the code here's what I can gather:

**cmake_pip**:
calls cmake from setup.py

pros:
- provides the usual `python setup.py install` interface, which makes it compatible with pip
- the project's CMakeLists.txt files don't need to know anything about cmake-pip or setup.py.  this means that if you don't care about pip/distutils stuff, you can just use cmake.

cons:
- additional dependency on cmake-pip.  setuptools provides the `setup_requires` option to install dependencies required for setup, but google is showing that it may not be as easy to use as it seems.  we can always embed cmake_pip into usdQt if that becomes a problem.

**bloerg.net**
calls setup.py from cmake (basically uses setup.py to install files in a standard fashion)

pros:
- no dependencies

cons:
- not compatible with pip

On paper, I prefer cmake-pip.  To get started I think we should just get working cmake files committed, then we can easily experiment with both of these approaches at a later date.

**Comment by pxmkuruc** [comment]
Got it. Let me look at getting cmake-pip installed.

File: usdQt/hierarchyModel.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+        """Retrieve the root of the current hierarchy model"""
+        rootProxy = self.__index.GetRoot()
+        return rootProxy.GetPrim() if rootProxy and not rootProxy.expired else None
+
+    def __OnObjectsChanged(self, notice, sender):
+        resyncedPaths = notice.GetResyncedPaths()
+        resyncedPaths = [path for path in resyncedPaths if path.IsPrimPath()]
+
+        if len(resyncedPaths) > 0:
+            with self.__LayoutChangedContext(self):
```
Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: My hesitation is needing to invalidate the models when an exception is thrown. It'll need to be a try/except/finally.

Comment (chadrik) [inline usdQt/hierarchyModel.py]: Yes, exactly.

**Comment by chadrik** [comment]
If you're pressed for time, I think the cmake-pip stuff can wait till later, after the cmake build is working.  Up to you.

File: usdQt/hierarchyModel.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+    def _GetPrimForIndex(self, modelIndex):
+        """Retrieve the prim for the input modelIndex
+
+        External clients should use UsdQt.roles.HierarchyPrimRole to access
+        the prim for an index"""
+        if modelIndex.isValid():
+            proxy = modelIndex.internalPointer()
+            proxy = compatability.ResolveValue(proxy)
+            if type(proxy) == _HierarchyCache.Proxy and not proxy.expired:
```
Comment (chadrik) [inline usdQt/hierarchyModel.py]: slight nit: `type(proxy) is _HierarchyCache.Proxy` is more correct here.  Although, unless your goal is to exclude subclasses of `Proxy`, it's probably more accurate (and idiomatic) to use `isinstance(proxy, _HierarchyCache.Proxy)`

Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: Fixed.  I don't know if we want subclasses of the Proxies just yet.  I think we'll have to go down this path at some point... but not quite yet  (For example, I expect we'll want to allow for specialization of handling of things like Primvars which might be spanned across multiple attributes if they are indexed).

**Comment by chadrik** [comment]
Success!  I got the extension built and importing using both setup.py and cmake (I fell back to cmake as a sanity check when I started running into problems).  I had to fix a few issues:

- added `PXR_NAMESPACE_OPEN/CLOSE_SCOPE` in all non-wrap*.cpp files
- added `PXR_NAMESSPACE_USING_DIRECTIVE` in wrap*.cpp files
- fixed an issue where primFilterCache was still using old debug.h

```diff
diff --git a/usdQt/debugCodes.cpp b/usdQt/debugCodes.cpp
index e14b6e7..73b4889 100644
--- a/usdQt/debugCodes.cpp
+++ b/usdQt/debugCodes.cpp
@@ -24,6 +24,8 @@
 
 #include "debugCodes.h"
 
+PXR_NAMESPACE_OPEN_SCOPE
+
 TF_REGISTRY_FUNCTION(TfDebug) {
     TF_DEBUG_ENVIRONMENT_SYMBOL(
         USDQT_DEBUG_HIERARCHYCACHE,
@@ -37,3 +39,5 @@ TF_REGISTRY_FUNCTION(TfDebug) {
     TF_DEBUG_ENVIRONMENT_SYMBOL(USDQT_DEBUG_UNDOSTACK,
                                 "Enabling Undo Stack debugging.");
 }
+
+PXR_NAMESPACE_CLOSE_SCOPE
diff --git a/usdQt/debugCodes.h b/usdQt/debugCodes.h
index cda063e..38a7116 100644
--- a/usdQt/debugCodes.h
+++ b/usdQt/debugCodes.h
@@ -27,7 +27,11 @@
 
 #include "pxr/base/tf/debug.h"
 
+PXR_NAMESPACE_OPEN_SCOPE
+
 TF_DEBUG_CODES(USDQT_DEBUG_HIERARCHYCACHE, USDQT_DEBUG_PRIMFILTERCACHE,
                USDQT_DEBUG_UNDOSTATEDELEGATE, USDQT_DEBUG_UNDOSTACK);
 
+PXR_NAMESPACE_CLOSE_SCOPE
+
 #endif
diff --git a/usdQt/hierarchyCache.cpp b/usdQt/hierarchyCache.cpp
index 9505a6c..9376be5 100644
--- a/usdQt/hierarchyCache.cpp
+++ b/usdQt/hierarchyCache.cpp
@@ -28,6 +28,8 @@
 #include "debugCodes.h"
 #include "hierarchyCache.h"
 
+PXR_NAMESPACE_OPEN_SCOPE
+
 typedef TfWeakPtr<UsdQt_HierarchyCache::Proxy> ProxyPtr;
 typedef TfRefPtr<UsdQt_HierarchyCache::Proxy> ProxyRefPtr;
 
@@ -239,4 +241,6 @@ void UsdQt_HierarchyCache::DebugFullIndex() {
                   << " [child count]: " << it.second->_GetChildren().size()
                   << std::endl;
     }
-}
\ No newline at end of file
+}
+
+PXR_NAMESPACE_CLOSE_SCOPE
diff --git a/usdQt/hierarchyCache.h b/usdQt/hierarchyCache.h
index c969fc7..46fffe2 100644
--- a/usdQt/hierarchyCache.h
+++ b/usdQt/hierarchyCache.h
@@ -37,6 +37,8 @@
 #include "pxr/usd/usd/primFlags.h"
 #include "pxr/usd/usd/stage.h"
 
+PXR_NAMESPACE_OPEN_SCOPE
+
 /// \class UsdQt_HierarchyCache
 ///
 /// The HierarchyCache is an internal class to UsdQt and provides a mapping
@@ -129,4 +131,6 @@ public:
     void DebugFullIndex();
 };
 
+PXR_NAMESPACE_CLOSE_SCOPE
+
 #endif
diff --git a/usdQt/primFilterCache.cpp b/usdQt/primFilterCache.cpp
index fcac0c1..0afe9a8 100644
--- a/usdQt/primFilterCache.cpp
+++ b/usdQt/primFilterCache.cpp
@@ -22,6 +22,7 @@
 // language governing permissions and limitations under the Apache License.
 //
 
+#include "debugCodes.h"
 #include "primFilterCache.h"
 
 #include "pxr/base/tf/enum.h"
@@ -30,6 +31,8 @@
 #include "pxr/base/work/loops.h"
 #include "tbb/parallel_reduce.h"
 
+PXR_NAMESPACE_OPEN_SCOPE
+
 TF_REGISTRY_FUNCTION(TfEnum) {
     TF_ADD_ENUM_NAME(UsdQtPrimFilterCache::Accept);
     TF_ADD_ENUM_NAME(UsdQtPrimFilterCache::Intermediate);
@@ -167,4 +170,6 @@ UsdQtPrimFilterCache::State UsdQtPrimFilterPathContains::operator()(
     if (not prim.GetChildren().empty())
         return UsdQtPrimFilterCache::Intermediate;
     return UsdQtPrimFilterCache::Reject;
-}
\ No newline at end of file
+}
+
+PXR_NAMESPACE_CLOSE_SCOPE
\ No newline at end of file
diff --git a/usdQt/primFilterCache.h b/usdQt/primFilterCache.h
index b1b9e06..511ce93 100644
--- a/usdQt/primFilterCache.h
+++ b/usdQt/primFilterCache.h
@@ -35,7 +35,7 @@
 
 #include "tbb/concurrent_unordered_map.h"
 
-#include "debug.h"
+PXR_NAMESPACE_OPEN_SCOPE
 
 /// \class UsdQtPrimFilterCache
 ///
@@ -123,4 +123,6 @@ public:
     UsdQtPrimFilterCache::State operator()(const UsdPrim& prim);
 };
 
+PXR_NAMESPACE_CLOSE_SCOPE
+
 #endif
\ No newline at end of file
diff --git a/usdQt/wrapHierarchyCache.cpp b/usdQt/wrapHierarchyCache.cpp
index aa8c694..7e768d7 100644
--- a/usdQt/wrapHierarchyCache.cpp
+++ b/usdQt/wrapHierarchyCache.cpp
@@ -31,6 +31,8 @@
 
 using namespace boost::python;
 
+PXR_NAMESPACE_USING_DIRECTIVE
+
 void wrapHierarchyCache() {
     {
         typedef UsdQt_HierarchyCache This;
diff --git a/usdQt/wrapPrimFilterCache.cpp b/usdQt/wrapPrimFilterCache.cpp
index 7ae05c9..106037f 100644
--- a/usdQt/wrapPrimFilterCache.cpp
+++ b/usdQt/wrapPrimFilterCache.cpp
@@ -39,6 +39,8 @@
 
 using namespace boost::python;
 
+PXR_NAMESPACE_USING_DIRECTIVE
+
 void wrapPrimFilterCache() {
     {
         typedef UsdQtPrimFilterCache This;
```

File: usdQt/hierarchyModel.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+            with self.__LayoutChangedContext(self):
+                persistentIndices = self.persistentIndexList()
+                indexToPath = {}
+                for index in persistentIndices:
+                    indexProxy = index.internalPointer()
+                    indexPrim = indexProxy.GetPrim()
+                    indexPath = indexPrim.GetPath()
+
+                    for resyncedPath in resyncedPaths:
+                        commonPath = resyncedPath.GetCommonPrefix(path)
```
Comment (chadrik) [inline usdQt/hierarchyModel.py]: the `path` variable here refers to the last value in `[path for path in resyncedPaths if path.IsPrimPath()]` on line 142, which I don't think is the intended behavior.

File: usdQt/hierarchyModel.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+            prim = self._GetPrimForIndex(modelIndex)
+            specifier = prim.GetSpecifier()
+            primType = prim.GetTypeName()
+            documentation = prim.GetDocumentation()
+            if specifier == Sdf.SpecifierDef:
+                toolTipString = "Defined %s"
+            elif specifier == Sdf.SpecifierOver:
+                toolTipString = "Undefined %s"
+            elif specifier == Sdf.SpecifierClass:
+                toolTipString = "Abstract %s"
```
Comment (chadrik) [inline usdQt/hierarchyModel.py]: I don't know offhand if these are the only specifiers allowed, but even if that's the case for the sake of clarity, and to a lesser degree forward-compatibility, we should probably have an `else` here to either raise an error (`Exception("this should never happen!)`) or provide a fallback `toolTipString = "Unknown %s"`.

File: usdQt/hierarchyModel.py (hunk 4 above)
Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: Thank you for catching that.  It wouldn't surprise me if in general we didn't tickle this, but I'll think of a way to add this to the regression tests.

File: usdQt/hierarchyModel.py (hunk 5 above)
Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: Good note. I just fixed that.

File: usdQt/hierarchyModel.py [hunk 6]
```diff
@@ -151,7 +151,7 @@ def __OnObjectsChanged(self, notice, sender):
                     indexPath = indexPrim.GetPath()
 
                     for resyncedPath in resyncedPaths:
-                        commonPath = resyncedPath.GetCommonPrefix(path)
+                        commonPath = resyncedPath.GetCommonPrefix(resyncedPath)
```
Comment (chadrik) [inline usdQt/hierarchyModel.py]: Is it meaningful for a path to get a common prefix with itself?  what's the result?

**Comment by chadrik** [comment]
Looks like all the notes have been resolved.  I had one more question, then I want to make sure I can compile this on my end, then we'll merge it and move on to the next batch.

File: usdQt/hierarchyModel.py (hunk 6 above)
Comment (pxmkuruc) [inline usdQt/hierarchyModel.py]: No, in my haste I picked the wrong path.  It should be indexPath.  Let me try and get a better test case that'll catch this...

Comment (chadrik) [reply usdQt/hierarchyModel.py]: Hey Matt, if you can get this last bit addressed we can get this merged in.

Comment (pxmkuruc) [reply usdQt/hierarchyModel.py]: I'll look at this today.  I think it's fixed locally.

**Comment by chadrik** [comment]
Looks great!

**Pull request merged**